src/libraries/Common/src/Interop/OSX/Swift.Runtime/CryptoKit.cs

```diff
@@ -0,0 +1,274 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Runtime.InteropServices.Swift;
+
+#pragma warning disable CS3016 // Arrays as attribute arguments are not CLS Compliant
+#pragma warning disable SYSLIB1051
+
+namespace Swift.Runtime
+{
+    /// <summary>
+    /// Represents ChaChaPoly in C#.
+    /// </summary>
+    internal unsafe partial struct ChaChaPoly
+    {
+        /// <summary>
+        /// Represents Nonce in C#.
+        /// </summary>
+        internal sealed unsafe partial class Nonce
+        {
+            private const int _payloadSize = 16;
+            internal void* payload;
+
+            internal Nonce()
+            {
+                payload = Marshal.AllocHGlobal(_payloadSize).ToPointer();
+                SwiftIndirectResult swiftIndirectResult = new SwiftIndirectResult(payload);
+                CryptoKit.PInvoke_ChaChaPoly_Nonce_Init(swiftIndirectResult);
+            }
+
+            internal Nonce(Data data)
+            {
+                payload = Marshal.AllocHGlobal(_payloadSize).ToPointer();
+                SwiftIndirectResult swiftIndirectResult = new SwiftIndirectResult(payload);
+                void* dataPtr = &data;
+
+                CryptoKit.PInvoke_ChaChaPoly_Nonce_Init2(swiftIndirectResult, dataPtr, data.Metadata(), default(DataProtocol).WitnessTable(data), out SwiftError error);
+
+                if (error.Value != null) // Should throw when nonce is not 12 bytes
+                {
+                    throw new Exception();
+                }
+            }
+
+            ~Nonce()
+            {
+                Marshal.FreeHGlobal(new IntPtr(payload));
```

Comment:
There is a conflict between convenience and performance. This was discussed in https://github.com/dotnet/designs/pull/312#discussion_r1517825410 .

It may be ok to depend on finalizers to cleanup for the app-developer facing Swift interop where convenience is preferred even if it leaves some performance on the table.

It is not necessarily what we want to do in this repo since we typically prefer performance over convenience for the internal implementation details here.

BTW: Other interops have similar conflict. For example, the app-developer facing classic COM interop depends on finalizers too. There are high-performance variants of COM interop that are less convenient to use and that have better performance: https://github.com/terrafx/terrafx.interop.windows/blob/main/sources/Interop/Windows/ComPtr%601.cs .

Reply:
Thanks for pointing this out.I've added IDisposable for non-frozen types in CryptoKit.